File: apps/backend/src/pets/pets.service.ts
import {
  ForbiddenException,
  Injectable,
  NotFoundException,
  UnprocessableEntityException,
} from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { StorageService } from '../storage/storage.service';
import { CreatePetDto } from './dto/create-pet.dto';
import { extname } from 'path';

@Injectable()
export class PetsService {
  constructor(private prisma: PrismaService, private storageService: StorageService) {}

  async create(dto: CreatePetDto, ownerId: string) {
    return this.prisma.pet.create({
      data: {
        ...dto,
        birthDate: dto.birthDate ? new Date(dto.birthDate) : undefined,
        ownerId,
      },
      include: { photos: true },
    });
  }

  async update(petId: string, dto: Partial<CreatePetDto>, userId: string) {
    const pet = await this.prisma.pet.findUnique({ where: { id: petId } });
    if (!pet) throw new NotFoundException('PET_NOT_FOUND');
    if (pet.ownerId !== userId) throw new ForbiddenException();
    return this.prisma.pet.update({
      where: { id: petId },
      data: {
        ...dto,
        birthDate: dto.birthDate ? new Date(dto.birthDate) : undefined,
      },
      include: { photos: true },
    });
  }

  async findOne(petId: string) {
    const pet = await this.prisma.pet.findUnique({
      where: { id: petId },
      include: { photos: { orderBy: { sortOrder: 'asc' } } },
    });
    if (!pet) throw new NotFoundException('PET_NOT_FOUND');
    return pet;
  }

  async addPhoto(petId: string, userId: string, file: Express.Multer.File, kind: string, sortOrder: number) {
    const pet = await this.prisma.pet.findUnique({ where: { id: petId }, include: { photos: true } });
    if (!pet) throw new NotFoundException('PET_NOT_FOUND');
    if (pet.ownerId !== userId) throw new ForbiddenException();
    if (pet.photos.length >= 5) throw new UnprocessableEntityException('MAX_PHOTOS_REACHED');

    const key = `pets/${petId}/${Date.now()}${extname(file.originalname)}`;
    const url = await this.storageService.uploadBuffer(file.buffer, key, file.mimetype);
    return this.prisma.photo.create({
      data: { petId, url, kind: kind as any, sortOrder },
    });
  }

  async deletePhoto(petId: string, photoId: string, userId: string) {
    const pet = await this.prisma.pet.findUnique({ where: { id: petId } });
    if (!pet) throw new NotFoundException('PET_NOT_FOUND');
    if (pet.ownerId !== userId) throw new ForbiddenException();

    const photo = await this.prisma.photo.findUnique({ where: { id: photoId } });
    if (!photo) throw new NotFoundException('PHOTO_NOT_FOUND');

    const key = this.storageService.keyFromUrl(photo.url);
    if (key) await this.storageService.deleteObject(key);

    return this.prisma.photo.delete({ where: { id: photoId } });
  }
}

File: apps/backend-java/src/main/java/com/pawpals/backend/pet/PetsService.java
package com.pawpals.backend.pet;

import com.pawpals.backend.common.ApiException;
import com.pawpals.backend.config.AppProperties;
import com.pawpals.backend.pet.dto.CreatePetRequest;
import com.pawpals.backend.pet.dto.PetResponse;
import com.pawpals.backend.pet.dto.PhotoResponse;
import com.pawpals.backend.pet.dto.UpdatePetRequest;
import com.pawpals.backend.user.UserRepository;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Set;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

@Service
@RequiredArgsConstructor
public class PetsService {

  private static final Set<String> ALLOWED_TYPES = Set.of("image/jpeg", "image/png");
  private static final int MAX_PHOTOS = 5;

  private final PetRepository petRepository;
  private final PhotoRepository photoRepository;
  private final UserRepository userRepository;
  private final AppProperties appProperties;

  public PetResponse create(CreatePetRequest dto, String ownerId) {
    Pet pet = new Pet();
    pet.setOwner(userRepository.getReferenceById(ownerId));
    pet.setName(dto.name());
    pet.setBreed(dto.breed());
    pet.setBio(dto.bio());
    pet.setTags(dto.tags());
    pet.setBirthDate(dto.birthDate());
    pet = petRepository.save(pet);
    return PetResponse.from(pet, List.of());
  }

  public PetResponse update(String petId, UpdatePetRequest dto, String userId) {
    Pet pet = requireOwnedPet(petId, userId);
    if (dto.name() != null) pet.setName(dto.name());
    if (dto.breed() != null) pet.setBreed(dto.breed());
    if (dto.bio() != null) pet.setBio(dto.bio());
    if (dto.tags() != null) pet.setTags(dto.tags());
    if (dto.birthDate() != null) pet.setBirthDate(dto.birthDate());
    pet = petRepository.save(pet);
    return PetResponse.from(pet, photosOf(petId));
  }

  public PetResponse findOne(String petId) {
    Pet pet = petRepository.findById(petId).orElseThrow(() -> ApiException.notFound("PET_NOT_FOUND"));
    return PetResponse.from(pet, photosOf(petId));
  }

  public PhotoResponse addPhoto(String petId, String userId, MultipartFile file, String kind, int sortOrder) {
    Pet pet = requireOwnedPet(petId, userId);
    if (photoRepository.countByPet_Id(pet.getId()) >= MAX_PHOTOS) {
      throw ApiException.unprocessable("MAX_PHOTOS_REACHED");
    }
    if (file == null || file.isEmpty()) {
      throw ApiException.unprocessable("FILE_REQUIRED");
    }
    if (!ALLOWED_TYPES.contains(file.getContentType())) {
      throw ApiException.unprocessable("INVALID_FILE_TYPE");
    }

    String ext = extensionOf(file.getOriginalFilename());
    String filename = System.currentTimeMillis() + ext;
    Path dir = Path.of(appProperties.uploads().dir(), "pets", petId).toAbsolutePath();
    try {
      Files.createDirectories(dir);
      file.transferTo(dir.resolve(filename));
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }

    Photo photo = new Photo();
    photo.setPet(pet);
    photo.setUrl("/uploads/pets/" + petId + "/" + filename);
    photo.setKind(PhotoKind.valueOf(kind));
    photo.setSortOrder(sortOrder);
    return PhotoResponse.from(photoRepository.save(photo));
  }

  public void deletePhoto(String petId, String photoId, String userId) {
    requireOwnedPet(petId, userId);
    Photo photo = photoRepository.findById(photoId).orElseThrow(() -> ApiException.notFound("PHOTO_NOT_FOUND"));

    Path filePath = Path.of(".", photo.getUrl()).normalize().toAbsolutePath();
    try {
      Files.deleteIfExists(filePath);
    } catch (IOException ignored) {
      // best-effort cleanup, matches Nest's fs.existsSync + unlinkSync guard
    }
    photoRepository.delete(photo);
  }

  private Pet requireOwnedPet(String petId, String userId) {
    Pet pet = petRepository.findById(petId).orElseThrow(() -> ApiException.notFound("PET_NOT_FOUND"));
    if (!pet.getOwnerId().equals(userId)) {
      throw ApiException.forbidden("FORBIDDEN");
    }
    return pet;
  }

  private List<PhotoResponse> photosOf(String petId) {
    return photoRepository.findByPet_IdOrderBySortOrderAsc(petId).stream().map(PhotoResponse::from).toList();
  }

  private String extensionOf(String originalFilename) {
    if (originalFilename == null) return "";
    int dot = originalFilename.lastIndexOf('.');
    return dot >= 0 ? originalFilename.substring(dot) : "";
  }
}
